docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,212 @@
++ string         sender     = 10;
+}
+```
+
+It is entirely up the application in determining how these fields are populated
+and with what values, e.g. the `sender` could be the signer and fee payer 
+of the transaction, the `priority` could be the cumulative sum of the fee(s).
+
+Only `sender` is required, while `priority` can be omitted which would result in
+using the default value of zero.
+
+### Mempool
+
+The existing concurrent-safe linked-list will be removed entirely in favor of a
+thread-safe map of `<sender:*Tx>`, i.e a mapping from `sender` to a single `*Tx`
+object, where each `*Tx` is the next valid and processable transaction from the
+given `sender`.
+
+On top of this mapping, we index all transactions by priority using a thread-safe
+priority queue, i.e. a [max heap](https://en.wikipedia.org/wiki/Min-max_heap).
+When a proposer is ready to select transactions for the next block proposal,
+transactions are selected from this priority index by highest priority order.
+When a transaction is selected and reaped, it is removed from this index and
+from the `<sender:*Tx>` mapping.
+
+We define `Tx` as the following data structure:
+
+```go
+type Tx struct {
+  // Tx represents the raw binary transaction data.
+  Tx []byte
+
+  // Priority defines the transaction's priority as specified by the application
+  // in the ResponseCheckTx response.
+  Priority int64
+
+  // Sender defines the transaction's sender as specified by the application in
+  // the ResponseCheckTx response.
+  Sender string
+
+  // Index defines the current index in the priority queue index. Note, if
+  // multiple Tx indexes are needed, this field will be removed and each Tx
+  // index will have its own wrapped Tx type.
+  Index int
+}
+```
+
+### Eviction
+
+Upon successfully executing `CheckTx` for a new `Tx` and the mempool is currently
+full, we must check if there exists a `Tx` of lower priority that can be evicted
+to make room for the new `Tx` with higher priority and with sufficient size
+capacity left.
+
+If such a `Tx` exists, we find it by obtaining a read lock and sorting the
+priority queue index. Once sorted, we find the first `Tx` with lower priority and
+size such that the new `Tx` would fit within the mempool's size limit. We then
+remove this `Tx` from the priority queue index as well as the `<sender:*Tx>`
+mapping.
+
+This will require additional `O(n)` space and `O(n*log(n))` runtime complexity.
+
+### Gossiping
+
+When gossiping new transactions to peers, we first obtain a read lock on the
+priority queue index. If we have not gossiped the top priority `Tx` for the
+given peer, we return this `Tx` as the next transaction to gossip. Otherwise,
+we sort the priority queue index and traverse in descending priority order until
+we find the first non-sent `Tx` to the peer.
+
+In order to facilitate this, we must also keep an ephemeral cache, mapping from
+`sender` to a list of unique `Tx` identifiers (e.g. the `Tx` hash). A `Tx`
```

Comment:
Is this the same `sender` as mentioned above or does it refer to a peer that the node is connected to?

Reply:
it's a cache of (sender, peerID, txID) tuples.

Reply:
Can you clarify this? Why is the sender involved here? Aren't we just trying to track which tx we sent to each peer?

Reply:
Oh its because when we have many txs per sender we have to gossip them in the right order per sender, regardless of priority? 

We should probably just start with support for more than 1 tx per sender since things like this might otherwise bite.